go/vt/mysqlctl/backupstats/stats.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backupstats
+
+import (
+	"time"
+
+	"vitess.io/vitess/go/stats"
+	vtstats "vitess.io/vitess/go/stats"
+)
+
+type Stats interface {
+	AddDuration(time.Duration)
+	SetDuration(time.Duration)
+	Scope(...Scope) Stats
+}
+
+type nopStats struct{}
+
+type scopedStats struct {
+	nsCountersWithLabels *vtstats.CountersWithMultiLabels
+	labels               []string
+}
+
+type unscopedStats struct {
+	sGauge               *vtstats.Gauge
+	nsCountersWithLabels *vtstats.CountersWithMultiLabels
+}
+
+func NopStats() Stats {
+	return &nopStats{}
+}
+
+func (ns *nopStats) Scope(...Scope) Stats {
+	return ns
+}
+
+func (ns *nopStats) AddDuration(d time.Duration) {}
+
+func (ns *nopStats) SetDuration(d time.Duration) {}
+
+func (uss *unscopedStats) AddDuration(d time.Duration) {
+	uss.sGauge.Add(int64(d.Seconds()))
+}
+
+func (uss *unscopedStats) SetDuration(d time.Duration) {
+	uss.sGauge.Set(int64(d.Seconds()))
+}
+
+func (uss *unscopedStats) Scope(scopes ...Scope) Stats {
+	return newScopedStats(uss.nsCountersWithLabels, nil, scopes...)
+}
+
+func newScopedStats(nsCountersWithLabels *stats.CountersWithMultiLabels, labels []string, scopes ...Scope) Stats {
+	if labels == nil {
+		labels = make([]string, 3)
+	}
+
+	for _, scope := range scopes {
+		labels[scope.Type] = scope.Value
```

Comment:
TODO add an out-of-bounds check. Probably fine to just log a warning and not panic, since the worst that can happen is screwed up metrics.  Alternately can return an `error` from the function and let that bubble up.